File: src/stores/history.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { useNotificationStore } from '@/stores/notification'
import request from '@/utils/request'

export type HistoryRunStatus = 'in_progress' | 'completed' | 'failed'
export type HistoryRunMode = 'tower' | 'daily_challenge'

export interface HistoryRunSummary {
  run_ref: string
  run_id: number
  mode: HistoryRunMode
  status: HistoryRunStatus
  day_key: string | null
  started_at: string | null
  ended_at: string | null
  highest_floor: number
  current_floor: number | null
  total_exp: number
  total_coins: number
  question_count: number
  last_hp: number | null
}

export interface HistoryAnswer {
  id: number
  answered_at: string | null
  correct: boolean | null
  answer_detail: Record<string, unknown> | null
  rating: number | null
  report: string | null
}

export interface HistoryQuestion {
  question_id: number
  order_index: number
  floor: number
  question_index: number | null
  level_id: number | null
  question_type: string | null
  content: Record<string, unknown> | null
  first_answer: HistoryAnswer | null
}

export interface HistoryRunDetail {
  run: HistoryRunSummary
  questions: HistoryQuestion[]
}

export const useHistoryStore = defineStore('history', () => {
  const notificationStore = useNotificationStore()

  const runs = ref<HistoryRunSummary[]>([])
  const currentRunDetail = ref<HistoryRunDetail | null>(null)
  const isLoadingRuns = ref(false)
  const isLoadingDetail = ref(false)

  async function fetchRuns() {
    isLoadingRuns.value = true
    try {
      const res = await request.get('/api/history/runs')
      if (res.data.success) {
        runs.value = res.data.data as HistoryRunSummary[]
      }
    } catch (error) {
      console.error('获取历史记录失败:', error)
      notificationStore.addNotification({
        title: '获取失败',
        description: '历史记录加载失败，请稍后重试',
        variant: 'destructive',
        duration: 3000,
      })
    } finally {
      isLoadingRuns.value = false
    }
  }

  async function fetchRunDetail(runRef: string) {
    isLoadingDetail.value = true
    currentRunDetail.value = null
    try {
      const res = await request.get(`/api/history/runs/${runRef}`)
      if (res.data.success) {
        currentRunDetail.value = res.data.data as HistoryRunDetail
      }
    } catch (error) {
      console.error('获取历史记录详情失败:', error)
      notificationStore.addNotification({
        title: '获取失败',
        description: '历史记录详情加载失败，请稍后重试',
        variant: 'destructive',
        duration: 3000,
      })
    } finally {
      isLoadingDetail.value = false
    }
  }

  function getRunSequenceNumber(runRef: string) {
    const towerRuns = runs.value.filter((run) => run.mode === 'tower')
    const index = towerRuns.findIndex((run) => run.run_ref === runRef)
    if (index === -1) {
      return null
    }
    return towerRuns.length - index
  }

  return {
    runs,
    currentRunDetail,
    isLoadingRuns,
    isLoadingDetail,
    fetchRuns,
    fetchRunDetail,
    getRunSequenceNumber,
  }
})

File: app/models/history.py
from datetime import datetime
from typing import Any, Optional

from sqlmodel import SQLModel


class HistoryAnswerRead(SQLModel):
    id: int
    answered_at: Optional[datetime] = None
    correct: Optional[bool] = None
    answer_detail: Optional[dict[str, Any]] = None
    rating: Optional[int] = None
    report: Optional[str] = None


class HistoryQuestionRead(SQLModel):
    question_id: int
    order_index: int
    floor: int
    question_index: Optional[int] = None
    level_id: Optional[int] = None
    question_type: Optional[str] = None
    content: Optional[dict[str, Any]] = None
    first_answer: Optional[HistoryAnswerRead] = None


class HistoryRunSummary(SQLModel):
    run_ref: str
    run_id: int
    mode: str
    status: str
    day_key: Optional[str] = None
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    highest_floor: int = 0
    current_floor: Optional[int] = None
    total_exp: int = 0
    total_coins: int = 0
    question_count: int = 0
    last_hp: Optional[int] = None


class HistoryRunDetail(SQLModel):
    run: HistoryRunSummary
    questions: list[HistoryQuestionRead]
